rwlib/source/fonts/Unicode.hpp

```diff
@@ -0,0 +1,303 @@
+ * @return number of bytes written
+ */
+size_t unicode_to_utf8(unicode_t unicode, char c[4]);
+
+/**
+ * @brief Iterate over a utf8 string stream. Output unicode data points.
+ */
+class Utf8UnicodeIterator {
+private:
+    /**
+     * @brief m_is Pointer to the utf8 stream to iterate over.
+     */
+    std::istream *m_is;
+
+    /**
+     * @brief m_finished true if the stream is finished/invalid.
+     */
+    bool m_finished;
+
+    /**
+     * @brief m_unicode Current unicode point.
+     */
+    unicode_t m_unicode;
+
+    /**
+     * @brief next_unicode Move to the next unicode point.
+     */
+    void next_unicode();
+
+public:
+    /**
+     * @brief Create an empty unicode iterator. The iterator is not good.
+     */
+    Utf8UnicodeIterator();
+    /**
+     * @brief Create a unicode iterator that iterates over a utf8 stream.
+     * @param is utf8 stream
+     */
+    Utf8UnicodeIterator(std::istream &is);
+
+    /**
+     * @brief operator ++ Move to the next unicode character.
+     * Only call this function when the iterator is good.
+     * @return this object
+     */
+    Utf8UnicodeIterator &operator ++();
+    /**
+     * @brief operator != Returns true if this iterator is good
+     * @return true if this iterator is good
+     */
+    bool operator!=(const Utf8UnicodeIterator &);
+
+    /**
+     * @brief operator * Returns the current unicode point
+     * ONly call this function when the iterator is good.
+     * @return unicode point
+     */
+    unicode_t operator *() const;
+
+    /**
+     * @brief unicode Returns the current unicode point
+     * Only call this function when the iterator is good.
+     * @return unicode point
+     */
+    unicode_t unicode() const;
+
+    /**
+     * @brief good Checks whether this stream is good.
+     * A stream is good when a unicode point is available.
+     * @return true if the stream is good, else false.
+     */
+    bool good() const;
+};
+
+/**
+ * @brief This class lets you iterate over a utf8 string using for ranged loop
+ */
+class Utf8UnicodeIteratorWrapper {
+public:
+    /**
+     * @brief Create a new wraper wrapper.
+     * @param s The string that the utf8 to Unicode iterator should iterate over
+     */
+    Utf8UnicodeIteratorWrapper(const std::string &s);
+
+    /**
+     * @brief Return the iterator at the start. This one returns unicode points.
+     * @return Utf8UnicodeIterator of which will read the unicode points.
+     */
+    Utf8UnicodeIterator begin();
+    /**
+     * @brief Return the unicode iterator at the end. This one is not good.
+     * No unicode points can be read from this object.
+     * @return Utf8UnicodeIterator Unicode iterator that is not good.
+     */
+    Utf8UnicodeIterator end();
+private:
+    /**
+     * @brief iss This utf8 stream holds the wrapped string and returns bytes to convert to unicode.
+     */
+    std::istringstream iss;
+};
+
+/**
+ * Unicode data points used by OpenRW.
+     */
+    enum UnicodeValue {
```

Comment:
enum class?

Reply:
Enum classes would make UnicodeValue strongly-typed. Implicitly converting a UnicodeValue to a unicode_t becomes impossible and requires a static_cast. https://en.wikipedia.org/wiki/C%2B%2B11#Strongly_typed_enumerations